onnxruntime/core/providers/qnn/builder/onnx_ctx_model_helper.cc

```diff
@@ -0,0 +1,124 @@
+                                 const std::unordered_map<std::string, OnnxTensorInfo>& inputs_info,
+                                 const std::vector<std::string>& output_names,
+                                 const std::unordered_map<std::string, OnnxTensorInfo>& outputs_info,
+                                 const std::string& model_description,
+                                 const std::string& sdk_build_version,
+                                 const std::string file_path,
+                                 unsigned char* buffer,
+                                 uint64_t buffer_size,
+                                 const logging::Logger& logger) {
+  std::unordered_map<std::string, int> domain_to_version = {{kOnnxDomain, 11}, {kMSDomain, 1}};
+  Model model(model_name, false, ModelMetaData(), PathString(), IOnnxRuntimeOpSchemaRegistryList(), domain_to_version,
+              {}, logger);
+  auto& graph = model.MainGraph();
+  graph.SetDescription(model_description);
+
+  onnxruntime::Version model_version = 5;
+  model.SetModelVersion(model_version);
+
+  using namespace ONNX_NAMESPACE;
+  std::vector<NodeArg*> inputs;
+  std::vector<NodeArg*> outputs;
+  ORT_RETURN_IF_ERROR(CreateNodeArgs(input_names, inputs_info, inputs, graph));
+  ORT_RETURN_IF_ERROR(CreateNodeArgs(output_names, outputs_info, outputs, graph));
+
+  auto& ep_node = graph.AddNode(graph_name,
+                                EPCONTEXT_OP,
+                                "Onnx Qnn context binary cache for graph partition: " + graph_name,
+                                inputs,
+                                outputs,
+                                nullptr,
+                                kMSDomain);
+
+  std::string cache_payload(buffer, buffer + buffer_size);
+  ep_node.AddAttribute("ep_engine_cache", cache_payload);
+  ep_node.AddAttribute("ep_sdk_version", sdk_build_version);
+  ep_node.AddAttribute("partition_name", graph_name);
+  ep_node.AddAttribute("source", kQnnExecutionProvider);
+
+  ORT_RETURN_IF_ERROR(graph.Resolve());
+
+  ORT_RETURN_IF_ERROR(Model::Save(model, file_path));
+
+  return Status::OK();
+}
+
+Status GetEpEngineCacheFromModel(const std::string& onnx_model_path,
+                                 std::string& ep_engine_cache,
+                                 const logging::Logger& logger) {
+  using namespace onnxruntime;
+  std::shared_ptr<Model> model;
+  ORT_RETURN_IF_ERROR(Model::Load(ToPathString(onnx_model_path), model, {}, logger));
+  const auto& graph = model->MainGraph();
+  ORT_RETURN_IF_ERROR(GetEpEngineCacheFromGraph(GraphViewer(graph), ep_engine_cache));
+
+  return Status::OK();
+}
+
+Status GetEpEngineCacheFromGraph(const onnxruntime::GraphViewer& graph_viewer, std::string& ep_engine_cache) {
+  const auto& node = graph_viewer.Nodes().begin();
+  NodeAttrHelper node_helper(*node);
+  ep_engine_cache = node_helper.Get("ep_engine_cache", ep_engine_cache);
+  return Status::OK();
```

Comment:
Since this function always returns `Status::OK()` should it just return the engine string instead?

Note that std::string's move constructor and move assignment operator should prevent redundant copies: 
```C++
std::string GetEpEngineCacheFromGraph(/*...*/) {
  // ...
  return node_helper.Get("ep_engine_cache", "");
}

ep_engine_cache = GetEpEngineCacheFromGraph(GraphViewer(graph));
```